homeassistant/components/ihc/__init__.py

```diff
@@ -0,0 +1,122 @@
+"""IHC platform."""
+# pylint: disable=too-many-arguments, too-many-instance-attributes
+import time
+import logging
+import os.path
+import xml.etree.ElementTree
+import voluptuous as vol
+
+import homeassistant.helpers.config_validation as cv
+from homeassistant.const import CONF_URL, CONF_USERNAME, CONF_PASSWORD
+from homeassistant.config import load_yaml_config_file
+
+from homeassistant.components.ihc import const
+
+REQUIREMENTS = ['ihcsdk==2.1.0']
+DOMAIN = 'ihc'
+
+CONF_INFO = 'info'
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_URL): cv.string,
+        vol.Required(CONF_USERNAME): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Required(CONF_INFO): cv.boolean
+    }),
+}, extra=vol.ALLOW_EXTRA)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def setup(hass, config):
+    """Setyp the IHC platform."""
+    from ihcsdk.ihccontroller import IHCController
+    url = config[DOMAIN].get(CONF_URL)
+    username = config[DOMAIN].get(CONF_USERNAME)
+    password = config[DOMAIN].get(CONF_PASSWORD)
+    ihc = IHCController(url, username, password)
+    ihc.info = config[DOMAIN].get(CONF_INFO)
+
+    if not ihc.authenticate():
+        _LOGGER.error("Unable to authenticate on ihc controller.")
+        return False
+
+    hass.data[DOMAIN] = IHCPlatform(ihc)
+
+    # Service functions
+
+    def set_runtime_value_bool(call):
+        """Set a IHC runtime bool value service function."""
+        ihcid = int(call.data.get('ihcid', 0))
+        value = bool(call.data.get('value', 0))
+        ihc.set_runtime_value_bool(ihcid, value)
+
+    def set_runtime_value_int(call):
+        """Set a IHC runtime integer value service function."""
+        ihcid = int(call.data.get('ihcid', 0))
+        value = int(call.data.get('value', 0))
+        ihc.set_runtime_value_int(ihcid, value)
+
+    def set_runtime_value_float(call):
+        """Set a IHC runtime float value service function."""
+        ihcid = int(call.data.get('ihcid', 0))
+        value = float(call.data.get('value', 0))
+        ihc.set_runtime_value_float(ihcid, value)
+
+    descriptions = load_yaml_config_file(
+        os.path.join(os.path.dirname(__file__), 'services.yaml'))
+
+    hass.services.register(DOMAIN, const.SERVICE_SET_RUNTIME_VALUE_BOOL,
+                           set_runtime_value_bool,
+                           descriptions[const.SERVICE_SET_RUNTIME_VALUE_BOOL])
```

Comment:
Please add service validation schemas for the services. The validation should validate the passed service data dict.

Reply:
Done